tests/tests/vertex_formats/mod.rs

```diff
@@ -0,0 +1,393 @@
+//! Tests that vertex formats pass through to vertex shaders accurately.
+
+use std::num::NonZeroU64;
+
+use wgpu::util::{BufferInitDescriptor, DeviceExt};
+
+use wgpu_test::{gpu_test, FailureCase, GpuTestConfiguration, TestParameters, TestingContext};
+
+#[derive(Debug, Copy, Clone)]
+enum TestCase {
+    UnormsAndSnorms,
+    UintsAndSintsSmall,
+    UintsBig,
+    SintsBig,
+    Floats,
+    Unorm1010102,
+}
+
+struct Test<'a> {
+    case: TestCase,
+    entry_point: &'a str,
+    attributes: &'a [wgt::VertexAttribute],
+    input: &'a [u8],
+    checksums: &'a [f32],
+}
+
+async fn vertex_formats_all(ctx: TestingContext) {
+    let attributes_block_0 = &wgpu::vertex_attr_array![
+        0 => Unorm8x4,
+        1 => Unorm16x2,
+        2 => Unorm16x4,
+        3 => Snorm8x4,
+        4 => Snorm16x2,
+        5 => Snorm16x4,
+        6 => Unorm8x2,
+        7 => Snorm8x2,
+    ];
+
+    let attributes_block_1 = &wgpu::vertex_attr_array![
+        0 => Uint8x4,
+        1 => Uint16x2,
+        2 => Uint16x4,
+        3 => Sint8x4,
+        4 => Sint16x2,
+        5 => Sint16x4,
+        6 => Uint8x2,
+        7 => Sint8x2,
+    ];
+
+    let attributes_block_2 = &wgpu::vertex_attr_array![
+        0 => Uint32,
+        1 => Uint32x2,
+        2 => Uint32x3,
+        3 => Uint32x4,
+    ];
+
+    let attributes_block_3 = &wgpu::vertex_attr_array![
+        0 => Sint32,
+        1 => Sint32x2,
+        2 => Sint32x3,
+        3 => Sint32x4,
+    ];
+
+    let attributes_block_4 = &wgpu::vertex_attr_array![
+        0 => Float32,
+        1 => Float32x2,
+        2 => Float32x3,
+        3 => Float32x4,
+        // TODO(SHADER_F16)
+        /*
+        4 => Float16x2,
+        5 => Float16x4,
+        */
```

Comment:
These should actually work without the `f16` extension. On the shader side we can expect `vec2<f32>` and `vec4<f32>`.

Reply:
Yes, but then wgsl requires the attribute to be declared as `vec2<f32>` but the vertex format remains `Float16x2`, so the unpack functions try to assign a `half2` to a `float2`, which isn't allowed. Also, the vertex buffer then has to be padded out with empty bytes, which the unpack functions then have to ignore/skip. It seems better to wait until we have `f16` support in wgsl.

Reply:
The spec currently allows the mapping of `Float16xN` to `vecN<f32>`. These are the relevant sections: https://gpuweb.github.io/gpuweb/#vertex-processing & https://gpuweb.github.io/gpuweb/wgsl/#channel-formats.

And since we don't implement `f16` yet, the unpacking functions just return `f32`.

https://github.com/gfx-rs/wgpu/blob/aeb2067e8120c1ff480625c00b9571db8d01d5a4/naga/src/back/msl/writer.rs#L4151-L4163

Reply:
I'll try to do this, but I've hit snags that I'm still working through. Some of the issues:

1) Regardless of the type used in the shader, the `Float16x2` format in the vertex buffer is 4 bytes, not 8 bytes.
2) If the `location(n)` annotation is declared as `vec2<f32>`, then the shader is expecting 8 bytes for that vec2 value.
3) When the test fills out the vertex buffer, it has to provide all those bytes to make the offsets work out.
4) If the test writes extra bytes (zeroes?) then the unpack function has to be revised to accept and ignore those bytes.
5) Presuming I can get all of this to work, are these the same steps that end-users have to do to use `Float16x2` vertex format? If so, where are we going to document that?

Perhaps I'm misunderstanding what needs to be done here, but this seems like a real mess. I'll see what I can get working.

Reply:
Huh. This time when I tried to make all the changes, it Just Worked. I guess my statement #3 above, is incorrect. Supplying 4 bytes for `Float16x2` works just fine.

Reply:
Yup, that is how it's supposed to work. `Float16x2` is 4 bytes in the vertex buffer and when we read them out we "expand" them into a `vec2<f32>` (8 bytes).

Reply:
I think the only changes need to get the new tests passing is to tweak the unpacking functions.